frame/im-online/src/lib.rs

```diff
@@ -333,14 +345,21 @@ pub mod pallet {
 	/// The current set of keys that may issue a heartbeat.
 	#[pallet::storage]
 	#[pallet::getter(fn keys)]
-	pub(crate) type Keys<T: Config> = StorageValue<_, Vec<T::AuthorityId>, ValueQuery>;
+	pub(crate) type Keys<T: Config> =
+		StorageValue<_, WeakBoundedVec<T::AuthorityId, T::MaxKeys>, ValueQuery>;
 
 	/// For each session index, we keep a mapping of `AuthIndex` to
 	/// `offchain::OpaqueNetworkState`.
 	#[pallet::storage]
 	#[pallet::getter(fn received_heartbeats)]
-	pub(crate) type ReceivedHeartbeats<T> =
-		StorageDoubleMap<_, Twox64Concat, SessionIndex, Twox64Concat, AuthIndex, Vec<u8>>;
+	pub(crate) type ReceivedHeartbeats<T: Config> = StorageDoubleMap<
+		_,
+		Twox64Concat,
+		SessionIndex,
+		Twox64Concat,
+		AuthIndex,
+		WeakBoundedVec<u8, T::MaxKeys>,
```

Comment:
the value here was a `Vec<u8>` which contains the encoded network state (i.e. an encoded value of type `OpaqueNetworkState`.
So it is no bounded by the max number of keys. We have to find or ask for the bound somewhere else.
It seems to be bounded by the max number of external addresses that the node can be reached at. maybe we can bound it directly in `sp-core::offchain::OpaqueNetworkState` or (and it is easier) ask for the bound in a generic param.

Reply:
The main reason why I did not bound it in `sp-core::off chain::OpaqueNetworkState` was because `WeakBoundedVec` is in `frame-support` and `sp-core` should not depend on it.

But indeed, you are right, I will have a think as to what is the best option to approach this, and will suggest it. Thanks!

Reply:
Question: `ReceivedHeartbeats` stores the encoding of the network state, but never seem to actually use that value, what is the purpose of storing that? Can we store its hash instead and that would naturally be bounded?

Reply:
Failing to find a better alternative, I have added MaxPeerInHeartbeats to bound ReceivedHeartbeats. It's a WeakBoundedVec and does not fail if the number is too big, but emits a warning. Let me know if this looks good to you.

Reply:
> Question: ReceivedHeartbeats stores the encoding of the network state, but never seem to actually use that value, what is the purpose of storing that? Can we store its hash instead and that would naturally be bounded?

Maybe it is only stored on chain so that it can be read, or can be used for some proof later, but I don't know either.

> Failing to find a better alternative, I have added MaxPeerInHeartbeats to bound ReceivedHeartbeats. It's a WeakBoundedVec and does not fail if the number is too big, but emits a warning. Let me know if this looks good to you.

this seems a good solution to me.